### PR TITLE
Validate password

### PR DESCRIPTION
**Pull request by AdamWills**
Added some basic password validation (addresses #3)

Have selected most rules to be disabled by default. NIST guidelines don't recommend any sort of password enforcement. Instead they recommend using a list of compromised passwords to compare against (which we are handling separately).

I'm not sure if we want to have different messages for different issues - but then we may want message() to allow returning an array of errors. Seems like it's something a little out of scope for this one though.

Source: https://pages.nist.gov/800-63-FAQ/#q-b06

### CODE REVIEW

File: src/Validation/Rules/Password.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+use Tempest\Validation\Rule;
+
+#[Attribute]
+final readonly class Password implements Rule
+{
+    private int $min;
+    private ?int $max;
+
+    public function __construct(
+        int $min = 8,
```
Comment (aidan-casey) [inline src/Validation/Rules/Password.php:17]: Despite NIST recommendations, let's revise this to be `12` as our default.

File: src/Validation/Rules/Password.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        ?int $max = null,
```
Comment (aidan-casey) [inline src/Validation/Rules/Password.php:18]: Let's remove `$max` as an option.

File: src/Validation/Rules/Password.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        private bool $mixedCase = false,
+        private bool $numbers = false,
+        private bool $letters = false,
+        private bool $symbols = false,
+    ) {
```
Comment (aidan-casey) [inline src/Validation/Rules/Password.php:23]: Let's add an option along the lines of `$restrictRepetitiveCharacters` (if you think of naming more user friendly, I'm open to it!)

File: src/Validation/Rules/Password.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        if ($this->symbols && ! preg_match('/\p{Z}|\p{S}|\p{P}/u', $value)) {
+            return false;
+        }
+
+        return true;
+    }
+
+    public function message(): string
+    {
+        return "Value should be a valid password";
```
Comment (aidan-casey) [inline src/Validation/Rules/Password.php:63]: Yeah... let's add more specific messages. 🙈

**Review by aidan-casey — CHANGES_REQUESTED** [review]
Thanks for the PR, @AdamWills! This is a big one to tackle.

I'd like to change just a few things if you don't mind reviewing!

File: src/Validation/Rules/Password.php (hunk 3 above)
Comment (AdamWills) [reply src/Validation/Rules/Password.php:23]: Can you define this in a little more detail?

File: src/Validation/Rules/Password.php (hunk 4 above)
Comment (AdamWills) [reply src/Validation/Rules/Password.php:63]: Perhaps for this PR, I'll return a single, more specific message. If we want to make a broader change to return multiple messages, let's separate that out.

File: src/Validation/Rules/Password.php (hunk 3 above)
Comment (aidan-casey) [reply src/Validation/Rules/Password.php:23]: Actually, ignore this for now. We can revisit later if we need to.

Was thinking things like "aaaaaa"

File: src/Validation/Rules/Password.php (hunk 4 above)
Comment (aidan-casey) [reply src/Validation/Rules/Password.php:63]: That's fine. We can strategize a bit for more complex scenarios like this.

File: src/Validation/Rules/Password.php (hunk 2 above)
Comment (AdamWills) [reply src/Validation/Rules/Password.php]: Done!

File: src/Validation/Rules/Password.php (hunk 4 above)
Comment (AdamWills) [reply src/Validation/Rules/Password.php]: Done!

**Comment by AdamWills** [comment]
Hm. Looks like there's failing tests on main...

**Comment by aidan-casey** [comment]
Thanks for the updates, @AdamWills! As far as failing tests go... I blame @brendt. 😉

**Pull request merged**